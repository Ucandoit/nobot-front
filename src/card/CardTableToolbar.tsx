import { makeStyles } from '@material-ui/core';
import IconButton from '@material-ui/core/IconButton';
import Toolbar from '@material-ui/core/Toolbar';
import Tooltip from '@material-ui/core/Tooltip';
import Typography from '@material-ui/core/Typography';
import FilterListIcon from '@material-ui/icons/FilterList';
import React from 'react';

const useStyles = makeStyles(theme => ({
  root: {
    paddingLeft: theme.spacing(2),
    paddingRight: theme.spacing(2)
  },
  title: {
    flex: '1 1 100%'
  }
}));

const CardTableToolbar = () => {
  const classes = useStyles();
  return (
    <Toolbar className={classes.root}>
      <Typography className={classes.title} variant="h6" id="cardTableTitle" component="div">
        Cards
      </Typography>
      <Tooltip title="Filter list">
        <IconButton aria-label="filter list">
          <FilterListIcon />
        </IconButton>
      </Tooltip>
    </Toolbar>
  );
};

export default CardTableToolbar;
