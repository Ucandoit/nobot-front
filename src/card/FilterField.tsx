import { FormControl, InputLabel, makeStyles, MenuItem, Select } from '@material-ui/core';
import React, { ChangeEvent, useCallback } from 'react';
import { Card, CardFilters } from '../helpers';

const useStyles = makeStyles(theme => ({
  filterField: {
    margin: theme.spacing(1),
    minWidth: '10rem'
  },
  select: {
    paddingTop: '1rem',
    paddingBottom: '1rem'
  }
}));

interface FilterFieldProps {
  property: keyof Card;
  label: string;
  items: (string | number)[];
  value: string | number | undefined;
  changeFilter: (property: keyof CardFilters, value: string) => void;
}

const FilterField = ({ property, label, items, value, changeFilter }: FilterFieldProps) => {
  const classes = useStyles();
  const onChange = useCallback(
    (event: ChangeEvent<{ value: unknown }>) =>
      changeFilter(property as keyof CardFilters, event.target.value as string),
    [property, changeFilter]
  );
  return (
    <FormControl variant="outlined" margin="dense" className={classes.filterField}>
      <InputLabel id={`${property}-label`}>{label}</InputLabel>
      <Select labelId={`${property}-label`} label={label} value={value || ''} onChange={onChange}>
        <MenuItem value="">-</MenuItem>
        {items.map(item => (
          <MenuItem key={item} value={item}>
            {item}
          </MenuItem>
        ))}
      </Select>
    </FormControl>
  );
};

export default FilterField;
