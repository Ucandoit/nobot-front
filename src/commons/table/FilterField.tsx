import { FormControl, InputLabel, makeStyles, MenuItem, Select } from '@material-ui/core';
import React, { ChangeEvent, useCallback } from 'react';
import { FilterType } from './types';

const useStyles = makeStyles(theme => ({
  filterField: {
    margin: theme.spacing(1),
    minWidth: '10rem'
  },
  select: {
    paddingTop: '1rem',
    paddingBottom: '1rem'
  }
}));

interface FilterFieldProps<T> {
  property: keyof T;
  label: string;
  type: FilterType;
  items?: Array<string | { key: string; value: string }>;
  value: string | undefined;
  changeFilter: (property: keyof T, value: string) => void;
}
const FilterField = <T,>({ property, label, type, items, value, changeFilter }: FilterFieldProps<T>) => {
  const classes = useStyles();
  const onChange = useCallback(
    (event: ChangeEvent<{ value: unknown }>) => changeFilter(property as keyof T, event.target.value as string),
    [property, changeFilter]
  );
  return (
    <FormControl variant="outlined" margin="dense" className={classes.filterField}>
      <InputLabel id={`${property}-label`}>{label}</InputLabel>
      {type === 'select' ? (
        <Select labelId={`${property}-label`} label={label} value={value || ''} onChange={onChange}>
          <MenuItem value="">-</MenuItem>
          {items &&
            items.map((item: string | { key: string; value: string }) => (
              <MenuItem
                key={typeof item === 'object' ? item.key : item}
                value={typeof item === 'object' ? item.value : item}
              >
                {item}
              </MenuItem>
            ))}
        </Select>
      ) : null}
    </FormControl>
  );
};

export default FilterField;
