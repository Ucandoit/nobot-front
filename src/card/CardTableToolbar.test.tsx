import { Button } from '@material-ui/core';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import CardTableToolbar from './CardTableToolbar';
import FilterField from './FilterField';

const changeFilter = jest.fn();
const resetFilter = jest.fn();

describe('CardTableToolbar', () => {
  let wrapper: ShallowWrapper;
  beforeEach(() => {
    wrapper = shallow(<CardTableToolbar filters={{}} changeFilter={changeFilter} resetFilter={resetFilter} />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with filters', () => {
    wrapper.setProps({ filters: { cost: 1.5, rarity: '煌' } });
    expect(wrapper).toMatchSnapshot();
  });

  it('should call changeFilter function on FilterField changeFilter triggers', () => {
    wrapper
      .find(FilterField)
      .at(0)
      .invoke('changeFilter')('rarity', '極');
    expect(changeFilter).toHaveBeenCalledTimes(1);
    expect(changeFilter).toHaveBeenCalledWith('rarity', '極');
  });

  it('should call resetFilter function on FilterField resetFilter triggers', () => {
    wrapper.find(Button).simulate('click');
    expect(resetFilter).toHaveBeenCalledTimes(1);
  });
});
