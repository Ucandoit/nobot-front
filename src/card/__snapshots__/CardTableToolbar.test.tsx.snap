// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardTableToolbar should render correctly 1`] = `
<WithStyles(ForwardRef(Toolbar))
  className="makeStyles-root-1"
>
  <FilterField
    changeFilter={[MockFunction]}
    items={
      Array [
        "煌",
        "極",
        "稀",
        "珍",
        "並",
        "宝",
        "誉",
      ]
    }
    label="Rarity"
    property="rarity"
  />
  <FilterField
    changeFilter={[MockFunction]}
    items={
      Array [
        0,
        1,
        2,
      ]
    }
    label="Star"
    property="star"
  />
  <FilterField
    changeFilter={[MockFunction]}
    items={
      Array [
        "火",
        "地",
        "風",
        "水",
        "空",
      ]
    }
    label="Property"
    property="property"
  />
  <FilterField
    changeFilter={[MockFunction]}
    items={
      Array [
        1,
        1.5,
        2,
        2.5,
        3,
        4,
      ]
    }
    label="Cost"
    property="cost"
  />
  <FilterField
    changeFilter={[MockFunction]}
    items={
      Array [
        "騎馬",
        "足軽",
        "鉄砲",
      ]
    }
    label="Military"
    property="military"
  />
  <FilterField
    changeFilter={[MockFunction]}
    items={
      Array [
        "武士",
        "軍師",
        "僧侶",
        "忍者",
        "剣豪",
        "傾奇",
        "姫",
        "茶人",
        "商人",
        "水軍",
      ]
    }
    label="Job"
    property="job"
  />
  <WithStyles(ForwardRef(Button))
    color="secondary"
    onClick={[MockFunction]}
    variant="contained"
  >
    Reset
  </WithStyles(ForwardRef(Button))>
</WithStyles(ForwardRef(Toolbar))>
`;

exports[`CardTableToolbar should render correctly with filters 1`] = `
<WithStyles(ForwardRef(Toolbar))
  className="makeStyles-root-1"
>
  <FilterField
    changeFilter={[MockFunction]}
    items={
      Array [
        "煌",
        "極",
        "稀",
        "珍",
        "並",
        "宝",
        "誉",
      ]
    }
    label="Rarity"
    property="rarity"
    value="煌"
  />
  <FilterField
    changeFilter={[MockFunction]}
    items={
      Array [
        0,
        1,
        2,
      ]
    }
    label="Star"
    property="star"
  />
  <FilterField
    changeFilter={[MockFunction]}
    items={
      Array [
        "火",
        "地",
        "風",
        "水",
        "空",
      ]
    }
    label="Property"
    property="property"
  />
  <FilterField
    changeFilter={[MockFunction]}
    items={
      Array [
        1,
        1.5,
        2,
        2.5,
        3,
        4,
      ]
    }
    label="Cost"
    property="cost"
    value={1.5}
  />
  <FilterField
    changeFilter={[MockFunction]}
    items={
      Array [
        "騎馬",
        "足軽",
        "鉄砲",
      ]
    }
    label="Military"
    property="military"
  />
  <FilterField
    changeFilter={[MockFunction]}
    items={
      Array [
        "武士",
        "軍師",
        "僧侶",
        "忍者",
        "剣豪",
        "傾奇",
        "姫",
        "茶人",
        "商人",
        "水軍",
      ]
    }
    label="Job"
    property="job"
  />
  <WithStyles(ForwardRef(Button))
    color="secondary"
    onClick={[MockFunction]}
    variant="contained"
  >
    Reset
  </WithStyles(ForwardRef(Button))>
</WithStyles(ForwardRef(Toolbar))>
`;
